### PR TITLE
App level Sharding Merge Request

### PR DESCRIPTION
**Pull request by amitkumar94**
Tests are not fixed with the changes, and i am adding new tests as well.

### CODE REVIEW

File: common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java [hunk 1]
```diff
@@ -13,43 +13,44 @@
...
  * Example:
  * {
- *     method1(); //call method1 which is annotated with transactional
+ * method1(); //call method1 which is annotated with transactional
  * }
  *
- * @Transactional
- * void method1() {
- *      method2(); //call method2 which is annotated with transactional
+ * @author shyam.akirala
```
Comment (shyamakirala) [inline common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java]: seems the author tag came in between

File: common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java [hunk 2]
```diff
@@ -64,27 +65,71 @@ public Object invoke(MethodInvocation invocation) throws Throwable {
...
+                                    sessionFactory = context.getRWSessionFactory(shardId);
+                                }
+                            }
+                        }
+                        catch (Exception ex){
+                            logger.error("Current Transactional Method doesn't have annotation @SelectDataSource Method_name:{} {}"
+                                    , invocation.getMethod().getName(), ex.getMessage());
+                            return new Error("Something wrong with Method's annotations");
+                        }
+                    }
```
Comment (shyamakirala) [inline common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java]: 'break' is missing, due to that sessionFactory gets overridden always

File: common/src/main/java/com/flipkart/flux/persistence/CryptHashGenerator.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    private static final String cryptHashAlgorithmPrefix = "SHA-256";
+
+    public static Character getUniformCryptHash(String stateMachineId) {
+        try {
+            MessageDigest md = MessageDigest.getInstance(cryptHashAlgorithmPrefix);
+            md.update(stateMachineId.getBytes());
+            String cryptHash = javax.xml.bind.DatatypeConverter.printHexBinary(md.digest()).toLowerCase();
+            return cryptHash.charAt(0);
+        } catch (Exception ex) {
+            throw new RuntimeException();
```
Comment (shyamakirala) [inline common/src/main/java/com/flipkart/flux/persistence/CryptHashGenerator.java]: empty runtime exception... difficult to figure out why it's failed. Wrap the actual exception with runtime exception.

File: common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java [hunk 4]
```diff
@@ -64,27 +65,71 @@ public Object invoke(MethodInvocation invocation) throws Throwable {
...
+            String shardKey;
+            ShardId shardId;
+            SessionFactory sessionFactory = null ;
+
+            try {
+                DataStorage dataStorage = invocation.getMethod().getAnnotation(DataStorage.class);
+                switch (dataStorage.value()) {
+                    case SHARDED: {
+                        try {
+                            SelectDataSource selectedDS = invocation.getMethod().getAnnotation(SelectDataSource.class);
```
Comment (shyamakirala) [inline common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java]: instead of having two annotations, it's good to have one, which gives both details DataSourceType and DataStorage

File: common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java [hunk 5]
```diff
@@ -64,27 +65,71 @@ public Object invoke(MethodInvocation invocation) throws Throwable {
...
+                                }
+                            }
+                        }
+                        catch (Exception ex){
+                            logger.error("Current Transactional Method doesn't have annotation @SelectDataSource Method_name:{} {}"
+                                    , invocation.getMethod().getName(), ex.getMessage());
+                            return new Error("Something wrong with Method's annotations");
+                        }
+                    }
+                    case REDRIVER: {
```
Comment (shyamakirala) [inline common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java]: REDRIVER -> NON_SHARDED ?

**Comment by shyamakirala** [comment]
will review once the tests are updated

File: common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java [hunk 6]
```diff
@@ -64,27 +65,71 @@ public Object invoke(MethodInvocation invocation) throws Throwable {
...
+                                    Object[] args = invocation.getArguments();
+                                    shardId = (ShardId) args[0];
+                                    sessionFactory = context.getROSessionFactory(shardId);
+                                }
+                                // in this case invocation method will provide shardKey i.e stateMachineId, as the first argument,
+                                // which will determine to which master shard the query should go to.
+                                case READ_WRITE: {
+                                    Object[] args = invocation.getArguments();
+                                    shardKey = (String) args[0];
+                                    Character shardHash = CryptHashGenerator.getUniformCryptHash(shardKey);
```
Comment (shyamakirala) [inline common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java]: Let's change it to string from character

File: common/src/main/java/com/flipkart/flux/persistence/CryptHashGenerator.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+public class CryptHashGenerator {
+
+    private static final String cryptHashAlgorithmPrefix = "SHA-256";
+
+    public static Character getUniformCryptHash(String stateMachineId) {
+        try {
+            MessageDigest md = MessageDigest.getInstance(cryptHashAlgorithmPrefix);
+            md.update(stateMachineId.getBytes());
+            String cryptHash = javax.xml.bind.DatatypeConverter.printHexBinary(md.digest()).toLowerCase();
+            return cryptHash.charAt(0);
```
Comment (shyamakirala) [inline common/src/main/java/com/flipkart/flux/persistence/CryptHashGenerator.java]: As discussed earlier, let's take first 2 characters

File: common/src/main/java/com/flipkart/flux/persistence/CryptHashGenerator.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package com.flipkart.flux.persistence;
```
Comment (shyamakirala) [inline common/src/main/java/com/flipkart/flux/persistence/CryptHashGenerator.java]: Add copyright comments, you can setup Intellij idea to to include copyright comments by default for all new files

File: common/src/main/java/com/flipkart/flux/persistence/CryptHashGenerator.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+import java.security.MessageDigest;
+
+/**
```
Comment (shyamakirala) [inline common/src/main/java/com/flipkart/flux/persistence/CryptHashGenerator.java]: Add class comment which says what the class does

File: common/src/main/java/com/flipkart/flux/persistence/DataStorage.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
+package com.flipkart.flux.persistence;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
```
Comment (shyamakirala) [inline common/src/main/java/com/flipkart/flux/persistence/DataStorage.java]: copyright and class comments

File: common/src/main/java/com/flipkart/flux/persistence/DataStorage.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+ * Created by amitkumar.o on 14/06/17.
+ */
+@Target(ElementType.METHOD)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface DataStorage {
```
Comment (shyamakirala) [inline common/src/main/java/com/flipkart/flux/persistence/DataStorage.java]: Append this functionality to SelectDataSource annotation

File: common/src/main/java/com/flipkart/flux/persistence/STORAGE.java [hunk 12]
```diff
@@ -0,0 +1,9 @@
+package com.flipkart.flux.persistence;
+
+/**
+ * Created by amitkumar.o on 14/06/17.
```
Comment (shyamakirala) [inline common/src/main/java/com/flipkart/flux/persistence/STORAGE.java]: copyright and class comments

File: runtime/src/main/resources/packaged/configuration.yml [hunk 13]
```diff
@@ -1,7 +1,10 @@
+flux.Shard.Config:
+  master.slave.pair.list: { "masterSlavePairList" : [ { "master":{ "shardId":1, "ip": "127.0.0.1", "shardKeys": ["0","1","2","3","4","5","6","7","8","9","a","b","c","d","e","f"] },"slave":{ "shardId":2, "ip":"127.0.0.1", "shardKeys": ["0","1","2","3","4","5","6","7","8","9","a","b","c","d","e","f"] }}  ]  }
```
Comment (shyamakirala) [inline runtime/src/main/resources/packaged/configuration.yml]: 1. Instead of keys, take range. 
2. This model looks redundant (shard keys and IDs) and error prone (mentioning master's shard ID 1 and slave's as 2)

ShardConfig: [
  {
    "ShardID": 1,
    "MasterIP": "localhost",
    "SlaveIP": "localhost",
    "StartOfRange": "00",
    "EndOfRange": "80"
  },
  {
    "ShardID": 2,
    "MasterIP": "10.20.30.40",
    "SlaveIP": "10.20.30.50",
    "StartOfRange": "80",
    "EndOfRange": "ff"
  }
]

Start of range is inclusive and End of range is exclusive
basically the keys which go in to first shard is 00 <= key < 80

File: runtime/src/main/resources/packaged/configuration.yml [hunk 14]
```diff
@@ -1,7 +1,10 @@
...
+
 flux.Hibernate:
   hibernate.dialect: "org.hibernate.dialect.MySQLDialect"
   hibernate.connection.driver_class: "com.mysql.jdbc.Driver"
-  hibernate.connection.url: "jdbc:mysql://localhost:3306/flux"
+  hibernate.connection.url: "jdbc:mysql://localhost:3306/flux_sharding"
```
Comment (shyamakirala) [inline runtime/src/main/resources/packaged/configuration.yml]: why are we having localhost in the url, the IP should be picked up from shard config

File: runtime/src/main/resources/packaged/configuration.yml [hunk 15]
```diff
@@ -12,7 +15,7 @@ flux.Hibernate:
 fluxReadOnly.Hibernate:
   hibernate.dialect: "org.hibernate.dialect.MySQLDialect"
   hibernate.connection.driver_class: "com.mysql.jdbc.Driver"
-  hibernate.connection.url: "jdbc:mysql://localhost:3306/flux"
+  hibernate.connection.url: "jdbc:mysql://localhost:3306/flux_sharding"
```
Comment (shyamakirala) [inline runtime/src/main/resources/packaged/configuration.yml]: same as above

File: runtime/src/main/resources/packaged/configuration.yml (hunk 13 above)
Comment (shyamakirala) [inline runtime/src/main/resources/packaged/configuration.yml]: Also we should be validating the cases like range overlaps, startOfRang < EndOfRange etc. and error out if there is something wrong

File: common/src/main/java/com/flipkart/flux/persistence/CryptHashGenerator.java (hunk 8 above)
Comment (amitkumar94) [inline common/src/main/java/com/flipkart/flux/persistence/CryptHashGenerator.java]: done

File: common/src/main/java/com/flipkart/flux/persistence/DataStorage.java (hunk 10 above)
Comment (amitkumar94) [inline common/src/main/java/com/flipkart/flux/persistence/DataStorage.java]: added

File: common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java (hunk 4 above)
Comment (amitkumar94) [inline common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java]: in order to make it one, developer need to give annotation value in array, which might lead to more mistakes, and is less readable.

File: common/src/main/java/com/flipkart/flux/persistence/CryptHashGenerator.java (hunk 3 above)
Comment (amitkumar94) [inline common/src/main/java/com/flipkart/flux/persistence/CryptHashGenerator.java]: corrected this one

File: runtime/src/main/resources/packaged/configuration.yml (hunk 14 above)
Comment (amitkumar94) [inline runtime/src/main/resources/packaged/configuration.yml]: This localhost is getting replaced while creating session factories with each host. This is just a default for how the url should look like

File: common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java [hunk 16]
```diff
@@ -64,50 +68,85 @@ public Object invoke(MethodInvocation invocation) throws Throwable {
...
+                                }
+                            }
+                        } catch (Exception ex) {
+                            logger.error("Current Transactional Method doesn't have annotation @SelectDataSource Method_name:{} {}"
+                                    , invocation.getMethod().getName(), ex.getStackTrace());
+                            return new Error("Something wrong with Method's annotations");
+                        }
+                        break;
+                    }
+                    case REDRIVER: {
```
Comment (shyamakirala) [inline common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java]: keep this as NON_SHARDED or SCHEDULER

File: common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java [hunk 17]
```diff
@@ -64,50 +68,85 @@ public Object invoke(MethodInvocation invocation) throws Throwable {
...
+                                    Object[] args = invocation.getArguments();
+                                    String shardKeyPrefix = CryptHashGenerator.getUniformCryptHash((String) args[0]);
+                                    sessionFactory = context.getRWSessionFactory(shardKeyPrefix);
+                                    break;
+                                }
+                            }
+                        } catch (Exception ex) {
+                            logger.error("Current Transactional Method doesn't have annotation @SelectDataSource Method_name:{} {}"
+                                    , invocation.getMethod().getName(), ex.getStackTrace());
+                            return new Error("Something wrong with Method's annotations");
```
Comment (shyamakirala) [inline common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java]: append ex.getMessage()

File: common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java [hunk 18]
```diff
@@ -64,50 +68,85 @@ public Object invoke(MethodInvocation invocation) throws Throwable {
...
+                    }
+                    case REDRIVER: {
+                        sessionFactory = context.getRedriverSessionFactory();
+                        break;
+                    }
+                }
+            } catch (Exception ex) {
+                logger.error("Current Transactional Method doesn't have annotation @DataStorage Method_name:{} {}"
+                        , invocation.getMethod().getName(), ex.getStackTrace());
+                return new Error("Something wrong with Method's annotations");
```
Comment (shyamakirala) [inline common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java]: append ex.getMessage()

File: common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java [hunk 19]
```diff
@@ -64,50 +68,85 @@ public Object invoke(MethodInvocation invocation) throws Throwable {
...
                 transaction.rollback();
-            }
-            throw e;
-        } finally {
-            if (transaction != null && session != null) {
-                ManagedSessionContext.unbind(context.getSessionFactory());
+                throw e;
+            } finally {
+                ManagedSessionContext.unbind(session.getSessionFactory());
+                logger.debug("Transaction completed for method : {} {} {}", invocation.getMethod().getName(), invocation.getMethod().getDeclaringClass());
```
Comment (shyamakirala) [inline common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java]: remove the extra {}

File: common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java [hunk 20]
```diff
@@ -64,50 +68,85 @@ public Object invoke(MethodInvocation invocation) throws Throwable {
...
                 session.close();
                 context.clear();
+                logger.debug("Clearing session from ThreadLocal Context {} {} {}", invocation.getMethod().getName(), invocation.getMethod().getDeclaringClass());
```
Comment (shyamakirala) [inline common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java]: remove the extra {}

File: common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java (hunk 19 above)
Comment (shyamakirala) [inline common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java]: null check is missing

File: common/src/main/java/com/flipkart/flux/persistence/DataStorage.java [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
```
Comment (shyamakirala) [inline common/src/main/java/com/flipkart/flux/persistence/DataStorage.java]: delete this interface.

File: examples/src/main/java/com/flipkart/flux/examples/concurrent/RunEmailMarketingWorkflow.java [hunk 22]
```diff
@@ -38,7 +42,11 @@ public static void main(String... args) throws Exception {
         final EmailMarketingWorkflow emailMarketingWorkflow = injector.getInstance(EmailMarketingWorkflow.class);
         /* Lets invoke our workflow */
         System.out.println("[Main] Starting workflow execution");
-        emailMarketingWorkflow.sendEmails(new Email("someBody","someone@flipkart.com"),new Email("someMore","someoneElse@fk.com"));
+        ExecutorService ex = Executors.newFixedThreadPool(100);
+        for(int i = 0 ; i < 10000; i++)
```
Comment (shyamakirala) [inline examples/src/main/java/com/flipkart/flux/examples/concurrent/RunEmailMarketingWorkflow.java]: remove the loop, execute only one

File: runtime/src/main/java/com/flipkart/flux/controller/WorkFlowExecutionController.java [hunk 23]
```diff
@@ -100,36 +118,39 @@ public WorkFlowExecutionController(EventsDAO eventsDAO, StateMachinesDAO stateMa
 
     /**
      * Retrieves the states which are dependant on this event and starts the execution of eligible states (whose all dependencies are met).
+     *
      * @param eventData
      * @param stateMachineInstanceId
      * @param correlationId
      */
-    public Set<State> postEvent(EventData eventData, Long stateMachineInstanceId, String correlationId) {
+    public Set<State> postEvent(EventData eventData, String stateMachineInstanceId, String correlationId) {
```
Comment (shyamakirala) [inline runtime/src/main/java/com/flipkart/flux/controller/WorkFlowExecutionController.java]: redundant IDs, as stateMachineInstanceId and correlationId are same, change the signature

File: runtime/src/main/java/com/flipkart/flux/controller/WorkFlowExecutionController.java [hunk 24]
```diff
@@ -100,36 +118,39 @@ public WorkFlowExecutionController(EventsDAO eventsDAO, StateMachinesDAO stateMa
...
         StateMachine stateMachine = null;
         if (stateMachineInstanceId != null) {
             stateMachine = retrieveStateMachine(stateMachineInstanceId);
-        } else if(correlationId != null) {
+        } else if (correlationId != null) {
             stateMachine = retrieveStateMachineByCorrelationId(correlationId);
```
Comment (shyamakirala) [inline runtime/src/main/java/com/flipkart/flux/controller/WorkFlowExecutionController.java]: all references to retrieveStateMachineByCorrelationId() should be replaced with retrieveStateMachine()

File: runtime/src/main/java/com/flipkart/flux/initializer/FluxInitializer.java [hunk 25]
```diff
@@ -55,56 +62,58 @@
...
                 break;
-            case "migrate" :
+            case "migrate":
                 if (args.length < 2) {
                     throw new RuntimeException("<migrate> must be followed with db name");
                 }
-                if (!(args[1].equals("flux") || args[1].equals("flux_redriver"))) {
-                    throw new RuntimeException("<migrate> works only for 'flux' or 'flux_redriver'");
+                if (!(args[1].equals("flux") || args[1].equals("flux_scheduler"))) {
+                    throw new RuntimeException("<migrate> works only for 'flux_sharding' or 'flux_redriver'");
```
Comment (shyamakirala) [inline runtime/src/main/java/com/flipkart/flux/initializer/FluxInitializer.java]: flux_sharding? it should be flux right

minor: change flux_redriver to flux_scheduler, seems this got missed in earlier

File: runtime/src/main/resources/packaged/configuration.yml [hunk 26]
```diff
@@ -1,35 +1,69 @@
+flux.Shard.Pair.Config: "{
```
Comment (shyamakirala) [inline runtime/src/main/resources/packaged/configuration.yml]: flux.Shard.Pair.Config: {
shardPairModelList: 
[
{...},
{...}
]
}

In the above model we can remove one level, we can model it as something like below

flux.Shard.Pair.Config: [
{...},
{...}
]

File: runtime/src/main/resources/packaged/configuration.yml [hunk 27]
```diff
@@ -1,35 +1,69 @@
...
+                                                                                   \"slaveIp\": \"10.32.105.84\",
+                                                                                   \"startKey\": \"c0\",
+                                                                                   \"endKey\":   \"ff\"
+                                                                             }
+                                                                          ]
+                                                  }"
 flux.Hibernate:
   hibernate.dialect: "org.hibernate.dialect.MySQLDialect"
   hibernate.connection.driver_class: "com.mysql.jdbc.Driver"
   hibernate.connection.url: "jdbc:mysql://localhost:3306/flux"
```
Comment (shyamakirala) [inline runtime/src/main/resources/packaged/configuration.yml]: we should remove "localhost" from here, it would be confusing

File: runtime/src/main/resources/packaged/configuration.yml [hunk 28]
```diff
@@ -1,35 +1,69 @@
+flux.Shard.Pair.Config: "{
+                                                    \"shardPairModelList\" :
+                                                                          [
+                                                                              {
+                                                                                \"shardId\": 1,
+                                                                                \"masterIp\": \"10.33.85.96\",
```
Comment (shyamakirala) [inline runtime/src/main/resources/packaged/configuration.yml]: IPs shouldn't be committed to public repos

File: runtime/src/main/resources/packaged/configuration.yml [hunk 29]
```diff
@@ -1,35 +1,69 @@
...
   hibernate.c3p0.min_size: 3
-  hibernate.c3p0.max_size: 10
+  hibernate.c3p0.max_size: 5
 
-flux_redriver.Hibernate:
+flux_scheduler.Hibernate:
   hibernate.dialect: "org.hibernate.dialect.MySQLDialect"
   hibernate.connection.driver_class: "com.mysql.jdbc.Driver"
-  hibernate.connection.url: "jdbc:mysql://localhost:3306/flux_redriver"
+  hibernate.connection.url: "jdbc:mysql://10.33.85.96:3306/flux_scheduler"
```
Comment (shyamakirala) [inline runtime/src/main/resources/packaged/configuration.yml]: remove IP

File: common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java (hunk 19 above)
Comment (amitkumar94) [inline common/src/main/java/com/flipkart/flux/guice/interceptor/TransactionInterceptor.java]: Intellij, was saying if( transaction != null) , is always true.

File: persistence-mysql/src/main/resources/flux/migrations/006_drop_correlatioId_from_StateMachines_table.sql [hunk 30]
```diff
@@ -0,0 +1,7 @@
+--liquibase formatted sql
+
+--changeset shyam.akirala:5 runOnChange:false
+
+ALTER TABLE `StateMachines` DROP COLUMN `correlationId` varchar(50) DEFAULT NULL;
```
Comment (amitkumar94) [inline persistence-mysql/src/main/resources/flux/migrations/006_drop_correlatioId_from_StateMachines_table.sql]: corrected

File: persistence-mysql/src/main/resources/flux/migrations/006_drop_correlatioId_from_StateMachines_table.sql [hunk 31]
```diff
@@ -0,0 +1,7 @@
...
+
+--rollback Alter Table `StateMachines` Drop Column `status`;
```
Comment (amitkumar94) [inline persistence-mysql/src/main/resources/flux/migrations/006_drop_correlatioId_from_StateMachines_table.sql]: corrected

File: runtime/src/main/java/com/flipkart/flux/controller/WorkFlowExecutionController.java (hunk 23 above)
Comment (amitkumar94) [inline runtime/src/main/java/com/flipkart/flux/controller/WorkFlowExecutionController.java]: corrected

File: runtime/src/main/java/com/flipkart/flux/initializer/FluxInitializer.java (hunk 25 above)
Comment (amitkumar94) [inline runtime/src/main/java/com/flipkart/flux/initializer/FluxInitializer.java]: corrected

File: runtime/src/main/java/com/flipkart/flux/dao/StatesDAOImpl.java [hunk 32]
```diff
@@ -16,99 +16,111 @@
...
     @Inject
-    public StatesDAOImpl(@Named("fluxSessionFactoryContext") SessionFactoryContext sessionFactoryContext) {
+    public StatesDAOImpl(@Named("fluxSessionFactoriesContext") SessionFactoryContext sessionFactoryContext) {
         super(sessionFactoryContext);
     }
 
     @Override
     @Transactional
+    @SelectDataSource(type = DataSourceType.READ_WRITE, storage = storage.SHARDED)
     public State create(State state) {
```
Comment (shyamakirala) [inline runtime/src/main/java/com/flipkart/flux/dao/StatesDAOImpl.java]: Actually this is a dead method, let's remove it

File: runtime/src/main/java/com/flipkart/flux/dao/StatesDAOImpl.java [hunk 33]
```diff
@@ -16,99 +16,111 @@
...
+    public void incrementRetryCount(String stateMachineId, Long stateId) {
         Query query = currentSession().createQuery("update State set attemptedNoOfRetries = attemptedNoOfRetries + 1 where id = :stateId and stateMachineId = :stateMachineId");
         query.setLong("stateId", stateId);
-        query.setLong("stateMachineId", stateMachineId);
+        query.setString("stateMachineId", stateMachineId);
         query.executeUpdate();
     }
 
+    /**
+     * Query should go to Default Shard , As it is a redriver Task
```
Comment (shyamakirala) [inline runtime/src/main/java/com/flipkart/flux/dao/StatesDAOImpl.java]: seems wrong comment

File: model/src/main/java/com/flipkart/flux/domain/State.java [hunk 34]
```diff
@@ -47,7 +47,7 @@
     /** Description for this State*/
     private String description;
     /** Id of the state machine to which this state belongs*/
-    private Long stateMachineId;
+    private String stateMachineId;
```
Comment (shyamakirala) [inline model/src/main/java/com/flipkart/flux/domain/State.java]: As discussed earlier, let's have <stateMachineId, taskId> as primary key for State table

File: model/src/main/java/com/flipkart/flux/domain/Event.java [hunk 35]
```diff
@@ -48,7 +48,7 @@
     private EventStatus status;
 
     /** Instance Id of state machine with which this event is associated */
-    private Long stateMachineInstanceId;
+    private String stateMachineInstanceId;
```
Comment (shyamakirala) [inline model/src/main/java/com/flipkart/flux/domain/Event.java]: Actually we can avoid autoincrement id for this table by having <stateMachineInstanceId, name> composite key

File: persistence-mysql/src/main/resources/flux/migrations/001_create_StateMachines_table.sql [hunk 36]
```diff
@@ -3,15 +3,14 @@
 --changeset shyam.akirala:1 runOnChange:false
 
 CREATE TABLE IF NOT EXISTS `StateMachines` (
-  `id` BIGINT NOT NULL AUTO_INCREMENT,
+  `id` VARCHAR(100) NOT NULL,
```
Comment (shyamakirala) [inline persistence-mysql/src/main/resources/flux/migrations/001_create_StateMachines_table.sql]: 100 -> 64 chars

File: persistence-mysql/src/main/resources/flux/migrations/002_create_States_table.sql [hunk 37]
```diff
@@ -8,7 +8,7 @@ CREATE TABLE IF NOT EXISTS `States` (
   `version` SMALLINT UNSIGNED NOT NULL,
   `description` VARCHAR(10) DEFAULT NULL,
   `dependencies` VARCHAR(1000) DEFAULT NULL,
-  `stateMachineId` BIGINT,
+  `stateMachineId` VARCHAR (100),
```
Comment (shyamakirala) [inline persistence-mysql/src/main/resources/flux/migrations/002_create_States_table.sql]: 64 chars

File: persistence-mysql/src/main/resources/flux/migrations/003_create_AuditRecords_table.sql [hunk 38]
```diff
@@ -4,7 +4,7 @@
 
 CREATE TABLE IF NOT EXISTS `AuditRecords` (
   `id` BIGINT NOT NULL AUTO_INCREMENT,
-  `stateMachineInstanceId` BIGINT NOT NULL,
+  `stateMachineInstanceId` VARCHAR(100) NOT NULL,
```
Comment (shyamakirala) [inline persistence-mysql/src/main/resources/flux/migrations/003_create_AuditRecords_table.sql]: 64 chars

File: persistence-mysql/src/main/resources/flux/migrations/004_create_Events_table.sql [hunk 39]
```diff
@@ -7,7 +7,7 @@ CREATE TABLE IF NOT EXISTS `Events` (
   `name` VARCHAR(255) NOT NULL,
   `type` VARCHAR(100) NOT NULL,
   `status` VARCHAR(50) DEFAULT NULL,
-  `stateMachineInstanceId` BIGINT NOT NULL,
+  `stateMachineInstanceId` VARCHAR(100) NOT NULL,
```
Comment (shyamakirala) [inline persistence-mysql/src/main/resources/flux/migrations/004_create_Events_table.sql]: 64 chars

File: persistence-mysql/src/main/resources/flux_scheduler/migrations/0001_create_message_table.sql [hunk 40]
```diff
@@ -4,6 +4,7 @@
 
 CREATE TABLE `ScheduledMessages` (
   `taskId` bigint(20) NOT NULL AUTO_INCREMENT,
+  `stateMachineId` VARCHAR (100) NOT NULL,
```
Comment (shyamakirala) [inline persistence-mysql/src/main/resources/flux_scheduler/migrations/0001_create_message_table.sql]: 64 chars

File: persistence-mysql/src/main/resources/flux_scheduler/migrations/0002_create_scheduled_events_table.sql [hunk 41]
```diff
@@ -0,0 +1,11 @@
+--liquibase formatted sql
+
+--changeset shyam.akirala:2 runOnChange:false
+
+CREATE TABLE `ScheduledEvents` (
+  `correlationId` VARCHAR(128) NOT NULL,
```
Comment (shyamakirala) [inline persistence-mysql/src/main/resources/flux_scheduler/migrations/0002_create_scheduled_events_table.sql]: 64 chars

File: runtime/src/main/java/com/flipkart/flux/dao/StatesDAOImpl.java [hunk 42]
```diff
@@ -16,99 +16,111 @@
...
     }
 
     @Override
     @Transactional
-    @SelectDataSource(DataSourceType.READ_ONLY)
-    public List findStatesByStatus(String stateMachineName, Timestamp fromTime, Timestamp toTime, String stateName, List<Status> statuses) {
+    @SelectDataSource(type = DataSourceType.READ_ONLY, storage = storage.SHARDED)
+    public List findStatesByStatus(ShardId shardId, String stateMachineName, Timestamp fromTime, Timestamp toTime, String stateName, List<Status> statuses) {
         Query query;
         String queryString = "select state.stateMachineId, state.id, state.status from StateMachine sm join sm.states state " +
```
Comment (shyamakirala) [inline runtime/src/main/java/com/flipkart/flux/dao/StatesDAOImpl.java]: this query returns wrong results, because the insertion order and timestamps are not related in case of string indexes.

Earlier the ids and timestamps are like
100 - t1
101 - t1
102 - t2
103 - t3

and now
abc - t1
xyz - t2 
def - t3

as random strings are inserted in to DB

will discus f2f and solve this

File: runtime/src/test/java/com/flipkart/flux/resource/StateMachineResourceTest.java [hunk 43]
```diff
@@ -139,101 +151,200 @@ public void testUnsideline() throws Exception {
...
     public void testFsmGraphCreation() throws Exception {
-        final StateMachine stateMachine = stateMachinePersistenceService.createStateMachine(objectMapper.readValue(this.getClass().getClassLoader().getResource("state_machine_definition_fork_join.json"), StateMachineDefinition.class));
+        final StateMachine stateMachine = stateMachinePersistenceService.createStateMachine(standardStateMachine, objectMapper.readValue(this.getClass().getClassLoader().getResource("state_machine_definition_fork_join.json"), StateMachineDefinition.class));
         final HttpResponse<String> stringHttpResponse = Unirest.get(STATE_MACHINE_RESOURCE_URL + "/" + stateMachine.getId() + "/fsmdata").header("Content-Type", "application/json").asString();
         assertThat(stringHttpResponse.getStatus()).isEqualTo(Response.Status.OK.getStatusCode());
         //TODO - we need a better assert here, but since we're using database IDs in the implementation, we cannot simply validate it with a static json
     }
 
-    @Test
+    //@Test
```
Comment (shyamakirala) [inline runtime/src/test/java/com/flipkart/flux/resource/StateMachineResourceTest.java]: commented out

File: scheduler/src/main/java/com/flipkart/flux/module/SchedulerModule.java [hunk 44]
```diff
@@ -78,19 +83,32 @@ public Configuration getConfiguration(YamlConfiguration yamlConfiguration) {
...
+
         configuration.addAnnotatedClass(ScheduledMessage.class);
+        configuration.addAnnotatedClass(ScheduledEvent.class);
+    }
+
+    @Provides
+    @Singleton
+    @Named("schedulerSessionFactoriesContext")
+    public SessionFactoryContext getSessionFactoryProvider(@Named("schedulerSessionFactory") SessionFactory schedulerSessionFactory) {
+        Map<ShardId, SessionFactory> fluxRWSessionFactoriesMap = new HashMap<ShardId, SessionFactory>();
```
Comment (shyamakirala) [inline scheduler/src/main/java/com/flipkart/flux/module/SchedulerModule.java]: null can be passed instead

File: runtime/src/main/resources/packaged/configuration.yml [hunk 45]
```diff
@@ -6,8 +18,9 @@ flux.Hibernate:
   hibernate.connection.password: ""
```
Comment (amitkumar94) [inline runtime/src/main/resources/packaged/configuration.yml]: This is needed to run Flux DB Migration
